Core/include/Acts/Navigation/NextNavigator.hpp

```diff
@@ -0,0 +1,357 @@
+  /// and keep thread-local navigation information
+  struct State : public NavigationState {
+    /// Navigation state - external state: the start surface
+    const Surface* startSurface = nullptr;
+    /// Navigation state - external state: the current surface
+    const Surface* currentSurface = nullptr;
+    /// Navigation state - external state: the target surface
+    const Surface* targetSurface = nullptr;
+    /// Indicator if the target is reached
+    bool targetReached = false;
+    /// Navigation state : a break has been detected
+    bool navigationBreak = false;
+
+    void reset(const GeometryContext& /*geoContext*/, const Vector3& /*pos*/,
+               const Vector3& /*dir*/, NavigationDirection /*navDir*/,
+               const Surface* /*ssurface*/, const Surface* /*tsurface*/) {
+      // Reset everything first
+      *this = State();
+
+      // TODO do stuff?
+    }
+  };
+
+  /// Constructor with configuration object
+  ///
+  /// @param cfg The navigator configuration
+  /// @param _logger a logger instance
+  explicit NextNavigator(Config cfg, std::shared_ptr<const Logger> _logger =
+                                         getDefaultLogger("NextNavigator",
+                                                          Logging::Level::INFO))
+      : m_cfg{cfg}, m_logger{std::move(_logger)} {}
+
+  /// @brief Navigator status call
+  ///
+  /// @tparam propagator_state_t is the type of Propagatgor state
+  /// @tparam stepper_t is the used type of the Stepper by the Propagator
+  ///
+  /// @param [in,out] state is the mutable propagator state object
+  /// @param [in] stepper Stepper in use
+  template <typename propagator_state_t, typename stepper_t>
+  void status(propagator_state_t& state, const stepper_t& stepper) const {
+    ACTS_VERBOSE(volInfo(state)
+                 << posInfo(state, stepper) << "Entering navigator::status.");
+
+    auto& nState = state.navigation;
+    fillNavigationState(state, stepper, nState);
+
+    if (inactive()) {
+      ACTS_VERBOSE("navigator inactive");
+      return;
+    }
+
+    if (nState.currentDetector == nullptr) {
+      initialize(state, stepper);
+      return;
+    }
+
+    if (nState.surfaceCandidate == nState.surfaceCandidates.end()) {
+      ACTS_VERBOSE("no surface candidates - waiting for target call");
+      return;
+    }
+
+    const Portal* nextPortal = nullptr;
+    const Surface* nextSurface = nullptr;
+    bool isPortal = false;
+    bool boundaryCheck = nState.surfaceCandidate->boundaryCheck;
+
+    if (nState.surfaceCandidate->surface != nullptr) {
+      nextSurface = nState.surfaceCandidate->surface;
+    } else if (nState.surfaceCandidate->portal != nullptr) {
+      nextPortal = nState.surfaceCandidate->portal;
+      nextSurface = &nextPortal->surface();
+      isPortal = true;
+    } else {
+      ACTS_ERROR("panic: not a surface not a portal - what is it?");
+      return;
+    }
+
+    // TODO not sure about the boundary check
```

Comment:
I think here wen need to understand:
- portals always have 'BoundaryCheck = true`, passive may at maximum have a tolerance, sensitives can be user defined, we need to sort of transport this information.

Reply:
Indeed, for the KalmanFilter, if you provide a measurement, you need to do 'BoundaryCheck = false', because it's the updater that should tell you if the hit is taken or not, not the surface bounds. We should keep this in mind and refine.